### ISSUE TITLE
Consumer offset in kafka topic

### ISSUE THREAD

**Issue by juise** [kafka4beam/brod#39]
Hi!

Would be amazing to have something like the high-level consumer that can use Kafka itself to manage offsets. Is it planned for the future?

**Comment by zmstone** [comment]
Hi !
We have a work in progress pr #37 which implements the consumer group feature.
Please have a try and let us know if this is what you are looking for, or if there is anything missing.

**Comment by juise** [comment]
Awesome! Looks like it's what I need, I'll try to make some benchmarks on this week. When do you plan merge this into master?

**Comment by juise** [comment]
Some times when I'm start publisher, send message or idle, I got:

> 2016-03-29 14:27:50.336 Undefined [error] [Undefined] CRASH REPORT Process <0.112.0> with 0 neighbours exited with reason: tcp_closed in brod_sock:handle_msg/3 line 208
> 2016-03-29 14:27:50.417 Undefined [error] [Undefined] client p_client1 metadata socket down localhost:9092
> Reason:tcp_closed
> 2016-03-29 14:27:56.153 Undefined [error] [Undefined] CRASH REPORT Process <0.135.0> with 0 neighbours exited with reason: tcp_closed in brod_sock:handle_msg/3 line 208
> 2016-03-29 14:27:56.153 Undefined [error] [Undefined] client c_client1 metadata socket down localhost:9092
> Reason:tcp_closed

where p_client1 - producer, c_client1 - consumer

**Comment by zmstone** [comment]
Hi
unfortunately we do not have solid time plan for now.

the tcp_closed 'error's on metadata sockets should perhaps have been 'info' instead,
or the log content should've been more verbose about the implication.

brod_client is 'lazy' on the metadata connections.
it tries to establish a connection when start/restart,
but only log and mark the death of it.
when the socket is dead, it will try to re-establish a new one when a new call to fetch metadata is made to the gen_server.

**Comment by juise** [comment]
Got it!

I found the another strange behavior, my code:

```
partitions(BootstrapHosts, Topic) ->
    case brod:start_link_client(BootstrapHosts, sys_info_client, []) of
        {ok, _ClientPid} ->
            ok;
        {error, {already_started, _ClientPid}} ->
            ok
    end,
    {ok, Partitions} = brod:get_partitions_count(sys_info_client, Topic),
    Partitions.

start_consumer(BootstrapHosts, Topic) ->
    Partitions = partitions(BootstrapHosts, Topic),
    [start_consumer(BootstrapHosts, Topic, N) || N <- lists:seq(0, Partitions - 1)].

start_consumer(BootstrapHosts, Topic, N) ->
    ConsumerGroupId = <<"consumer-group-", Topic/binary>>,
    ConsumerGroupConfig = [{offset_commit_policy, commit_to_kafka_v2}, {offset_commit_interval_seconds, 1}],
    ConsumerId = binary_to_atom(<<"consumer-", (integer_to_binary(N))/binary, "-", Topic/binary>>, utf8),
    {ok, _ClientPid} = brod:start_link_client(BootstrapHosts, ConsumerId, []),
    {ok, _Subscriber} = brod_group_subscriber:start_link(ConsumerId, ConsumerGroupId, [Topic], ConsumerGroupConfig, [], ?MODULE, []),
    {N, ConsumerId}.
```

and when I run:

```
140> kafka_bench:start_consumer([{"localhost", 9092}], <<"my-repl-topic-4">>).
[{0,'consumer-0-my-repl-topic-4'},
 {1,'consumer-1-my-repl-topic-4'},
 {2,'consumer-2-my-repl-topic-4'},
 {3,'consumer-3-my-repl-topic-4'}]
141> 2016-03-29 18:30:00.174 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=,generation=0,pid=<0.13553.0>):
connected to group coordinator host.ru:9092
2016-03-29 18:30:00.175 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=,generation=0,pid=<0.13561.0>):
connected to group coordinator host.ru:9092
2016-03-29 18:30:00.177 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-3c921395-ece0-476a-84bd-38f46c330ebb,generation=1,pid=
<0.13553.0>):
elected=true
2016-03-29 18:30:00.177 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=,generation=0,pid=<0.13557.0>):
connected to group coordinator host.ru:9092
2016-03-29 18:30:00.183 Undefined [error] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-3c921395-ece0-476a-84bd-38f46c330ebb,generation=1,pid
=<0.13553.0>):
failed to join group
reason:'RebalanceInProgress'
[{brod_group_controller,sync_group,1,[{file,"/Users/juise/Documents/kafka_bench/_build/default/lib/brod/src/brod_group_controller.erl"},{line,543}]},{brod_group_controller,do_stabilize,3,[{file,"/Users/juise/Documents/kafka_bench/_build/default/lib/brod/src/brod_group_controller.erl"},{line,438}]},{brod_group_controller,do_stabilize,3,[{file,"/Users/juise/Documents/kafka_bench/_bui
ld/default/lib/brod/src/brod_group_controller.erl"},{line,439}]},{brod_group_controller,handle_info,2,[{file,"/Users/juise/Documents/kafka_bench/_build/default/lib/brod/src/brod_group_controller.e
rl"},{line,269}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,593}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,659}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,2
37}]}]
2016-03-29 18:30:00.183 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-3c921395-ece0-476a-84bd-38f46c330ebb,generation=1,pid=
<0.13553.0>):
re-joining group, reason:'RebalanceInProgress'
2016-03-29 18:30:00.183 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=,generation=0,pid=<0.13568.0>):
connected to group coordinator host.ru:9092
2016-03-29 18:30:00.186 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-3c921395-ece0-476a-84bd-38f46c330ebb,generation=2,pid$
<0.13553.0>):
elected=true
2016-03-29 18:30:00.186 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-b1f22096-2607-4ebe-9833-b602850d5bb6,generation=2,pid$
<0.13561.0>):
elected=false
2016-03-29 18:30:00.186 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-cae3034d-6421-4dc8-867d-aff05ecafda2,generation=2,pid$
<0.13557.0>):
elected=false
2016-03-29 18:30:00.186 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-7a5ef9b4-5aad-49d2-b3fb-14a02a0823af,generation=2,pid$
<0.13568.0>):
elected=false
2016-03-29 18:30:00.189 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-cae3034d-6421-4dc8-867d-aff05ecafda2,generation=2,pid$
<0.13557.0>):
assignments:[{<<"my-repl-topic-4">>,[{partition_assignment,1,-1,<<>>}]}]
2016-03-29 18:30:00.189 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-3c921395-ece0-476a-84bd-38f46c330ebb,generation=2,pid$
<0.13553.0>):
assignments:[{<<"my-repl-topic-4">>,[{partition_assignment,2,-1,<<>>}]}]
2016-03-29 18:30:00.189 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-b1f22096-2607-4ebe-9833-b602850d5bb6,generation=2,pid$
<0.13561.0>):
assignments:[{<<"my-repl-topic-4">>,[{partition_assignment,3,-1,<<>>}]}]
2016-03-29 18:30:00.190 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-4,memberId=consumer-group-my-repl-topic-4-7a5ef9b4-5aad-49d2-b3fb-14a02a0823af,generation=2,pid=
<0.13568.0>):
assignments:[{<<"my-repl-topic-4">>,[{partition_assignment,0,-1,<<>>}]}]
2016-03-29 18:30:00.198 Undefined [info] [Undefined] client 'consumer-0-my-repl-topic-4' connected to host.ru:9092
2016-03-29 18:30:00.198 Undefined [info] [Undefined] client 'consumer-2-my-repl-topic-4' connected to host.ru:9092
2016-03-29 18:30:00.199 Undefined [info] [Undefined] client 'consumer-3-my-repl-topic-4' connected to host.ru:9092
2016-03-29 18:30:00.199 Undefined [info] [Undefined] client 'consumer-1-my-repl-topic-4' connected to host.ru:9092
```

**Comment by zmstone** [comment]
judging by the logs, I see it's perfectly normal.
is 'failed to join group' strange to you or?
'RebalanceInProgress' is what happens when group members join one after another.
as a matter of fact, there is no way to add ALL members to the group in one go.
i.e. if there are 4 members to be in the group, there will be at least 3 'RebalanceInProgress' exceptions to receive.
Other exceptions like IllegalGeneration etc may also happen during group state transition.

You need to worry only if the consumer group does not enter a stable state (where all members are active and have their assignments received) for long, or the exceptions keeps popping up as if in a dead loop.

edit: i was wrong about 'at least 3 RebalanceInProgress' exceptions' the number should be 1.
--- i.e. it will elect one as group leader, then the group leader will receive this exception, if lucky, all other 3 may join in the same generation, hence no need for more rebalance.

**Comment by zmstone** [comment]
btw, you don't need one client_id for each member.
you may start_link one client, and use it for all your consumers.
this will lessen the number of opened sockets to the brokers.

**Comment by juise** [comment]
Yep, exactly about "failed to join group" and others. I think will be better to hide this kind of errors from end user, and log them only in the debug mode?

I think is more efficient to have the different connections between each publishers and consumers, isn't it?

**Comment by zmstone** [comment]
good point. will consider add a DEBUG level log, not sure how to do it without adding a lager dependency though. will take it to the team had have a discussion on it.

if you think bidirectional traffic may affect performance,
you may start one client for producers and one for consumers.

would be interesting to see how big is the difference if use only one client for both.
we haven't done any benchmark tests so far.

**Comment by juise** [comment]
Yep, looks like lager need here )

I'll try to check the performance when:
- producers and consumers share the same connection
- producers share the one connection, and consumers the another connection
- producers and consumers have personal connection each

**Comment by juise** [comment]
Hi!

I find out new strange behavior, I have to consumer groups:
1. from erlang -> consumer-group-my-repl-topic-1
2. from ./kafka-console-consumer-group -> console-consumer-77906

and the output for kafka-consumer-groups.sh looks like it but or something like:

```
$ ./kafka-consumer-groups.sh --new-consumer --bootstrap-server localhost:9092 --describe --group console-consumer-77906
GROUP, TOPIC, PARTITION, CURRENT OFFSET, LOG END OFFSET, LAG, OWNER
console-consumer-77906, my-repl-topic-1, 0, 225009, 225009, 0, consumer-1_/127.0.0.1

$ ./kafka-consumer-groups.sh --new-consumer --bootstrap-server localhost:9092 --describe --group consumer-group-my-repl-topic-1
Error while executing consumer group command Error reading field 'user_data': java.nio.BufferUnderflowException
org.apache.kafka.common.protocol.types.SchemaException: Error reading field 'user_data': java.nio.BufferUnderflowException
        at org.apache.kafka.common.protocol.types.Schema.read(Schema.java:71)
        at org.apache.kafka.clients.consumer.internals.ConsumerProtocol.deserializeAssignment(ConsumerProtocol.java:109)
        at kafka.admin.AdminClient$$anonfun$describeConsumerGroup$1.apply(AdminClient.scala:161)
        at kafka.admin.AdminClient$$anonfun$describeConsumerGroup$1.apply(AdminClient.scala:160)
        at scala.collection.immutable.List.map(List.scala:273)
        at kafka.admin.AdminClient.describeConsumerGroup(AdminClient.scala:160)
        at kafka.admin.ConsumerGroupCommand$KafkaConsumerGroupService.describeGroup(ConsumerGroupCommand.scala:314)
        at kafka.admin.ConsumerGroupCommand$ConsumerGroupService$class.describe(ConsumerGroupCommand.scala:84)
        at kafka.admin.ConsumerGroupCommand$KafkaConsumerGroupService.describe(ConsumerGroupCommand.scala:302)
        at kafka.admin.ConsumerGroupCommand$.main(ConsumerGroupCommand.scala:63)
        at kafka.admin.ConsumerGroupCommand.main(ConsumerGroupCommand.scala)
```

**Comment by zmstone** [comment]
Hi
we are probably not following the tooling convention of the user_data format.
will investigate.

**Comment by juise** [comment]
I'm sorry, but one more question: I start producer, produce message, start consumer and it doesn't consume messages. I stop consumer, stop produces, produce more messages, start consumer and it's start to consume messages. What I'm doing wrong?

**Comment by zmstone** [comment]
i'm afraid i have to ask for more context or code to understand the problem.
normally producers and consumers should run independently.

**Comment by juise** [comment]
Sure, here full gist of my module - https://gist.github.com/juise/fcafbd76e4ac704c7bef6174f30995f5

Then, I do:

```
> ./kafka-topics.sh --zookeeper localhost:2181 --create --topic=my-repl-topic-8 --partitions 1 --replication-factor 1

1> kafka_bench:producer_perf([{"localhost", 9092}], <<"my-repl-topic-8">>, 100, 100).
2016-03-30 14:18:08.399 Undefined [info] [Undefined] client 'producer-0-my-repl-topic-8' connected to host.ru:9092
2016-03-30 14:18:08.443 app@host.ru [info] [kafka_bench] Producer: 100 packets, 66.946 ms
ok

2>  kafka_bench:start_consumer([{"localhost", 9092}], <<"my-repl-topic-8">>).
[{0,'consumer-0-my-repl-topic-8'}]
(app@host.ru)3> 2016-03-30 14:18:23.507 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-8,memberId=,generation=0,pid=<0.507.0>):
connected to group coordinator host.ru:9092
2016-03-30 14:18:23.508 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-8,memberId=consumer-group-my-repl-topic-8-dc5478ed-22e4-4fcb-8eb9-598f758433c0,generation=1,pid=<0.507.0>):
elected=true
2016-03-30 14:18:23.512 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-8,memberId=consumer-group-my-repl-topic-8-dc5478ed-22e4-4fcb-8eb9-598f758433c0,generation=1,pid=<0.507.0>):
assignments:[{<<"my-repl-topic-8">>,[{partition_assignment,0,-1,<<>>}]}]
2016-03-30 14:18:23.522 Undefined [info] [Undefined] client 'consumer-0-my-repl-topic-8' connected to host.ru:9092
```

And nothing happens!

**Comment by zmstone** [comment]
Hi, did not go through your code in details, make a quick guess here:
ok, so you start producer, and produce 100 messages, and producer is done, right ?
although you have started consumers seemingly in parallel, there might be a chance that the producer have already stopped producing messages.

the main issue is that subscribers by default subscribes to offset -1 (the latest), if you change it to -2, you might be able to see a difference.
and since you are committing offsets to kafka, when you restart your consumer with the same old group id, it should continue from the last committed offset.

**Comment by juise** [comment]
Looks like you are right, but how and where I should change offset from -1 to -2?

**Comment by zmstone** [comment]
at line 119, instead of [] for ConsumerConfig, use [{begin_offset, -2}] instead.

**Comment by juise** [comment]
I'm doing as you say, but unfortunately it's doesn't work, looks like somewhere mistake, I noticed that even I use [{begin_offset, -2}] in log I see:

```
2016-03-30 15:02:28.251 Undefined [info] [Undefined] group controller (groupId=consumer-group-my-repl-topic-11,memberId=consumer-group-my-repl-topic-11-02e2eda5-aead-43c3-9875-ba11bbe8f047,generation=1,pid=<0.2395.0>):
assignments:[{<<"my-repl-topic-11">>,[{partition_assignment,0,-1,<<>>}]}]
```

Once again, sorry for problems!

**Comment by zmstone** [comment]
ok, it's the broker which replies -1 when there is no offset commit history found for a given partition in the request.
in our code, we were expecting UnknownTopicOrPartition exception in this case (which was the case for version 0 requests).
I think we need to first understand better about the implication of -1 received from kafka, then perhaps come up with as solution and hopefully get it fixed in #38

**Comment by juise** [comment]
Got it, thanks!

**Comment by zmstone** [comment]
our pleasure, and thank you for the feedback.

**Comment by zmstone** [comment]
@juise just noticed that this is actually a documented behavior, just pushed the fix to #37 , now -2 should work.

**Comment by zmstone** [comment]
@juise the display issue has been fixed in the latest commit to #37 : 
https://github.com/klarna/brod/pull/37/commits/f8e7e80ffa557986d0894a969e46ceb3196a2b8f
